cpp/include/raft/linalg/map_then_reduce.cuh

```diff
@@ -76,9 +92,35 @@ void mapThenSumReduceImpl(Type *out, size_t len, MapOp map, cudaStream_t stream,
 template <typename Type, typename MapOp, int TPB = 256, typename... Args>
 void mapThenSumReduce(Type *out, size_t len, MapOp map, cudaStream_t stream,
                       const Type *in, Args... args) {
-  mapThenSumReduceImpl<Type, MapOp, TPB, Args...>(out, len, map, stream, in,
-                                                  args...);
+  mapThenReduceImpl<Type, MapOp, sum_tag, TPB, Args...>(
+    out, len, (Type)0, map, sum_tag(), stream, in, args...);
 }
 
+/**
+ * @brief CUDA version of map and then generic reduction operation
+ * @tparam Type data-type upon which the math operation will be performed
+ * @tparam MapOp the device-lambda performing the actual map operation
+ * @tparam ReduceLambda the device-lambda performing the actual reduction
+ * @tparam TPB threads-per-block in the final kernel launched
+ * @tparam Args additional parameters
+ * @param out the output reduced value (assumed to be a device pointer)
+ * @param len number of elements in the input array
+ * @param neutral The neutral element of the reduction operation. For example:
+ *    0 for sum, 1 for multiply, +Inf for Min, -Inf for Max
+ * @param map the device-lambda
+ * @param op the reduction device lambda
+ * @param stream cuda-stream where to launch this kernel
+ * @param in the input array
+ * @param args additional input arrays
+ */
+
+template <typename Type, typename MapOp, typename ReduceLambda, int TPB = 256,
+          typename... Args>
+void mapThenReduce(Type *out, size_t len, Type neutral, MapOp map,
```

Comment:
Maybe we should specialize the templates for input and output differently as `Type1` and `Type2`? Summation or multiplication reduction could easily cause an underflow/overflow, and this will help protect against that

Reply:
Thanks Divye, this is a good observation! There is a small issue here: the [CUB block reduce ops](https://nvlabs.github.io/cub/classcub_1_1_block_reduce.html#a089953b3bdfe7c48208632d0cc2ac1fb) that we use under the hood have the same input and output type. So there will be TPB = 256 elements reduced into the same Type1, and the rest of the atomic reductions we can perform with Type2. It probably still make sense to make the suggested change, I will push a commit soon.